koku/koku/trino_database.py

```diff
@@ -78,6 +84,130 @@ def query_id(self) -> t.Optional[str]:
         return self._trino_error.query_id
 
 
+class NoSuchKeyViolation:
+    """Detect NoSuchKey violation in an exception or error message."""
+
+    NO_SUCH_KEY_REGEX_STR = r"NoSuchKey"
+    NO_SUCH_KEY_REGEX = re.compile(NO_SUCH_KEY_REGEX_STR)
+
+    def __init__(self, _exception):
+        """Accepts Exception or str"""
+        self.message = _exception if isinstance(_exception, str) else str(_exception)
+        self.__is_no_such_key_violation = bool(self.NO_SUCH_KEY_REGEX.search(self.message))
+
+    @property
+    def is_no_such_key_violation(self):
+        """Returns bool indicating if NoSuchKey violation was detected."""
+        return self.__is_no_such_key_violation
+
+    def __bool__(self):
+        return self.__is_no_such_key_violation
+
+    def __repr__(self):
+        return str(self)
+
+    def __str__(self):
+        if self.__is_no_such_key_violation:
+            return f"NoSuchKey error detected: {self.message}"
+        else:
+            return "No NoSuchKey error detected."
+
+
+class TrinoNoSuchKeyException(Exception):
```

Comment:
We should probably include some attributes on this execption so that if it does get raised and passed to `log_json` the `str` will show useful information in the logs.